### PR TITLE
restructure inspect to handle polymorphic resources

### PR DESCRIPTION
**Pull request by juanvallejo**
cc @deads2k

### CODE REVIEW

File: pkg/cmd/inspect.go [hunk 1]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
 	return nil
 }
 
+func objectReferenceToResourceInfo(clientGetter genericclioptions.RESTClientGetter, ref *configv1.ObjectReference) (*resource.Info, error) {
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: I think it's time to make packages for commands.

File: pkg/cmd/inspect.go [hunk 2]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+	infos, err := b.Do().Infos()
+	if err != nil {
+		return nil, err
+	}
+
+	return infos[0], nil
+}
+
+func infoToContextKey(info *resource.Info) string {
+	return fmt.Sprintf("%s/%s.%s/%s/%s", info.Namespace, info.ResourceMapping().GroupVersionKind.Group, info.ResourceVersion, info.ResourceMapping().Resource.Resource, info.Name)
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: no resourceversion in this key

File: pkg/cmd/inspect.go [hunk 3]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+
+	infos, err := b.Do().Infos()
+	if err != nil {
+		return nil, err
+	}
+
+	return infos[0], nil
+}
+
+func infoToContextKey(info *resource.Info) string {
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: doc the meaning of this key

File: pkg/cmd/inspect.go [hunk 4]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+	return fmt.Sprintf("%s/%s.%s/%s/%s", info.Namespace, info.ResourceMapping().GroupVersionKind.Group, info.ResourceVersion, info.ResourceMapping().Resource.Resource, info.Name)
+}
+
+// collectPolymorphicResourceInfo receives an object to gather debugging data for, and a context to keep track of
+// already-seen objects when following related-object reference chains.
+func (o *InspectOptions) collectPolymorphicResourceInfo(info *resource.Info, context map[string]bool) ([]string, error) {
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: make context a struct that contains a "visited" string set

Comment (deads2k) [inline pkg/cmd/inspect.go]: `InspectResource`

File: pkg/cmd/inspect.go [hunk 5]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+	context[infoToContextKey(info)] = true
+
+	unstr, ok := info.Object.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type. Expecting %q but got %T", "*unstructured.Unstructured", info.Object)
+	}
+
+	skippedNamespaces := []string{}
+
+	switch info.ResourceMapping().Resource.Resource {
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: no, switch on teh groupresource

Comment (deads2k) [inline pkg/cmd/inspect.go]: switch on the actual info.Object()'s groupresource

File: pkg/cmd/inspect.go [hunk 6]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+	case "clusteroperators":
+		// save clusteroperator resources to disk
+		if err := o.gatherClusterOperatorResource(path.Join(o.baseDir, "/clusteroperator"), unstr); err != nil {
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: `base/cluster-scoped/config.openshift.io(group)/clusteroperators(resource)/<name>.yaml` for individuals and `base/cluster-scoped/config.openshift.io/clusteroperators(resource).yaml` for lists I think

Comment (deads2k) [inline pkg/cmd/inspect.go]: move the logic from this case into it's own function I think

File: pkg/cmd/inspect.go (hunk 4 above)
Comment (deads2k) [inline pkg/cmd/inspect.go]: structure like polymorphic subresources where it's a bare method not attached to the options struct

File: pkg/cmd/inspect.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/inspect.go]: have a variant `objectRefToContextKey` so you don't have to retrieve things you've already seen.

File: pkg/cmd/inspect.go [hunk 7]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+
+		// obtain associated objects for the current clusteroperator resources
+		relatedObjReferences, err := obtainClusterOperatorRelatedObjects(unstr)
+		if err != nil {
+			return nil, err
+		}
+
+		errs := []error{}
+		for _, relatedRef := range relatedObjReferences {
+			relatedInfo, err := objectReferenceToResourceInfo(o.configFlags, relatedRef)
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: intercept already visited items here too

File: pkg/cmd/inspect.go [hunk 8]
```diff
@@ -155,6 +155,93 @@ func (o *InspectOptions) Validate() error {
...
+		}
+
+		if len(errs) > 0 {
+			return skippedNamespaces, errors.NewAggregate(errs)
+		}
+	case "namespaces":
+		if err := o.gatherNamespaceData(path.Join(o.baseDir, "namespaces", info.Name), info.Name); err != nil {
+			if kapierrs.IsNotFound(err) {
+				skippedNamespaces = append(skippedNamespaces, info.Name)
+				return skippedNamespaces, err
```
Comment (deads2k) [inline pkg/cmd/inspect.go]: why?  We've already visited it, so there's no reason for a special case, right?

Comment (juanvallejo) [reply pkg/cmd/inspect.go]: This is for cases where the namespace is not found at all, so that we can list all of them (if any) at the end

**Comment by juanvallejo** [comment]
@deads2k thanks, comments addressed

File: pkg/cmd/inspect/inspect.go [hunk 9]
```diff
@@ -167,102 +167,134 @@ func (o *InspectOptions) Run() error {
...
 	errs := []error{}
 	if err := o.gatherConfigResourceData(path.Join(o.baseDir, "/cluster-scoped-resources/config.openshift.io")); err != nil {
 		errs = append(errs, err)
 	}
 
-	// gather operator.openshift.io resource data
+	// then, gather operator.openshift.io resource data
 	if err := o.gatherOperatorResourceData(path.Join(o.baseDir, "/cluster-scoped-resources/operator.openshift.io")); err != nil {
 		errs = append(errs, err)
 	}
```
[the remark this answers is not in the document]
Comment (davemulford) [reply pkg/cmd/inspect/inspect.go]: That makes sense to me. Having _sane_ defaults is always a good idea.

[the remark this answers is not in the document]
Comment (juanvallejo) [reply pkg/cmd/inspect/inspect.go]: Agree, cc @deads2k

File: pkg/cmd/inspect/inspect.go [hunk 10]
```diff
@@ -167,102 +167,134 @@ func (o *InspectOptions) Run() error {
...
 		}
+	}
 
-		namespaces, err := obtainClusterOperatorNamespaces(info.Object)
-		if err != nil {
-			errs = append(errs, err)
-			continue
+	if len(skippedNamespaces) > 0 {
+		for _, namespace := range skippedNamespaces {
+			log.Printf("Data collection skipped namespace %q. Unable to find namespace...\n", namespace)
```
[the remark this answers is not in the document]
Comment (juanvallejo) [reply pkg/cmd/inspect/inspect.go]: yes, makes sense

File: cmd/must-gather.go [hunk 11]
```diff
@@ -46,7 +43,7 @@ func NewCmdMustGather(streams genericclioptions.IOStreams) *cobra.Command {
 		},
 	}
 
-	cmd.AddCommand(mustgather.NewCmdInspect("openshift-must-gather", streams))
+	cmd.AddCommand(inspect.NewCmdInspect("openshift-must-gather", streams))
```
Comment (deads2k) [inline cmd/must-gather.go]: nit: passing a parent doesn't add us much value, does it?

File: pkg/cmd/inspect/inspect.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/cmd/inspect/inspect.go]: > Agree, cc @deads2k

I'm ok always adding those when collecting anything.

File: pkg/cmd/inspect/inspect.go [hunk 12]
```diff
@@ -167,102 +167,138 @@ func (o *InspectOptions) Run() error {
...
+		return errors.NewAggregate(allErrs)
+	}
+
+	log.Printf("Finished successfully with no errors.\n")
+	return nil
+}
+
+// InspectResource receives an object to gather debugging data for, and a context to keep track of
+// already-seen objects when following related-object reference chains.
+func InspectResource(info *resource.Info, context *resourceContext, o *InspectOptions) ([]string, error) {
```
Comment (deads2k) [inline pkg/cmd/inspect/inspect.go]: separate file please

File: pkg/cmd/inspect/inspect.go [hunk 13]
```diff
@@ -167,102 +167,138 @@ func (o *InspectOptions) Run() error {
...
+	if !ok {
+		return nil, fmt.Errorf("unexpected type. Expecting %q but got %T", "*unstructured.Unstructured", info.Object)
+	}
+
+	notFoundNamespaces := []string{}
+
+	switch info.ResourceMapping().Resource.GroupResource() {
+	case configv1.GroupVersion.WithResource("clusteroperators").GroupResource():
+		// save clusteroperator resources to disk
+		if err := o.gatherClusterOperatorResource(path.Join(o.baseDir, "/cluster-scoped-resources", "/"+unstr.GroupVersionKind().Group, "/clusteroperators"), unstr); err != nil {
```
Comment (deads2k) [inline pkg/cmd/inspect/inspect.go]: make this idiot proof.  Only pass the basedir, the rest is computed to always be the same.  Constant for `cluster-scoped-resources` please

File: pkg/cmd/inspect/inspect.go [hunk 14]
```diff
@@ -167,102 +167,138 @@ func (o *InspectOptions) Run() error {
...
-	}
 
-	if len(skippedNamespaces) > 0 {
-		for _, namespace := range skippedNamespaces {
-			log.Printf("Data collection skipped namespace %q. Unable to find namespace...\n", namespace)
+		if len(errs) > 0 {
+			return notFoundNamespaces, errors.NewAggregate(errs)
+		}
+	case corev1.SchemeGroupVersion.WithResource("namespaces").GroupResource():
+		if err := o.gatherNamespaceData(path.Join(o.baseDir, "namespaces", info.Name), info.Name); err != nil {
```
Comment (deads2k) [inline pkg/cmd/inspect/inspect.go]: same thing about paths here.  Also, namespaces are cluster scoped for this purpose. neat, right?

Make a special group folder called `core` for `""`

File: pkg/cmd/inspect/inspect.go [hunk 15]
```diff
@@ -167,102 +167,138 @@ func (o *InspectOptions) Run() error {
...
+			if kapierrs.IsNotFound(err) {
+				notFoundNamespaces = append(notFoundNamespaces, info.Name)
+				return notFoundNamespaces, err
+			}
+		}
+	default:
+		// save the current object to disk
+		filename := fmt.Sprintf("%s.yaml", unstr.GetName())
+		objPath := path.Join(o.baseDir, "/namespaces", "/"+unstr.GetNamespace(), "/"+info.ResourceMapping().Resource.Resource)
```
Comment (deads2k) [inline pkg/cmd/inspect/inspect.go]: choose paths based on the presence of namespace

**Comment by deads2k** [comment]
let's go for it

**Pull request merged**